mage_ai/streaming/sources/google_cloud_pubsub.py

```diff
@@ -0,0 +1,127 @@
+from google.api_core import retry
+from google.cloud import pubsub_v1
+from dataclasses import dataclass
+from mage_ai.shared.config import BaseConfig
+from mage_ai.streaming.constants import DEFAULT_BATCH_SIZE
+from mage_ai.streaming.sources.base import BaseSource
+from typing import Callable
+import os
+
+
+@dataclass
+class GoogleCloudPubSubConfig(BaseConfig):
+    project_id: str
+    topic_id: str
+    subscription_id: str
+    timeout: int = 5
+    batch_size: int = DEFAULT_BATCH_SIZE
+    pubsub_emulator_host: str = None  # e.g., "host.docker.internal:8085"
+    google_application_credentials: str = None  # e.g., "./google_credentials.json"
+
+
+class GoogleCloudPubSubSource(BaseSource):
+    """
+    Handles data transfer between a Google Cloud Pub/Sub topic and the Mage app.
+
+    GOOGLE_APPLICATION_CREDENTIALS environment could be used to set the Google Cloud
+    credentials file for authentication.
+    """
+
+    config_class = GoogleCloudPubSubConfig
+
+    def _exist_subscription(self, project_id: str) -> bool:
+        project_path = f"projects/{project_id}"
+        subscriptions = self.subscriber_client.list_subscriptions(
+            request={"project": project_path}
+        )
+        for subscription in subscriptions:
+            if subscription.name == self.subscription_path:
+                return True
+        return False
+
+    def _create_subscription(
+            self, project_id: str, topic_id: str, subscription_id: str) -> None:
+        """Create a new pull subscription on the given topic."""
+        # [START pubsub_create_pull_subscription]
+        if self._exist_subscription(project_id):
+            self._print(f"Subscription already exists: {self.subscription_path}")
+            return
+
+        publisher = pubsub_v1.PublisherClient()
+        topic_path = publisher.topic_path(project_id, topic_id)
+        subscription = self.subscriber_client.create_subscription(
+            request={"name": self.subscription_path, "topic": topic_path}
+        )
+        self._print(f"Subscription created: {subscription}")
+        # [END pubsub_create_pull_subscription]
+
+    def init_client(self) -> None:
+        if self.config.pubsub_emulator_host is not None:
+            os.environ["PUBSUB_EMULATOR_HOST"] = self.config.pubsub_emulator_host
+        if self.config.google_application_credentials is not None:
+            os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = \
```

Comment:
I would not suggest setting the credentials in environment variable. There could be conflicts between the streaming pipeline processes.

Reply:
Sure, changed the code to load the credentials file directly instead of relying on changing the environment variable.